chronos/asyncfutures2.nim

```diff
@@ -398,6 +402,49 @@ proc `cancelCallback=`*[T](future: Future[T], cb: CallbackFunc) =
   ## This callback will be called immediately as ``future.cancel()`` invoked.
   future.cancelcb = cb
 
+{.push stackTrace: off.}
+proc internalContinue[T](fut: pointer) {.gcsafe, raises: [Defect].}
+
+proc futureContinue*[T](fut: Future[T]) {.gcsafe, raises: [Defect].} =
+
+  # Used internally by async transformation
+  try:
+    if not(fut.closure.finished()):
+      var next = fut.closure(fut)
+      # Continue while the yielded future is already finished.
+      while (not next.isNil()) and next.finished():
+        next = fut.closure(fut)
+        if fut.closure.finished():
+          break
+
+      if next == nil:
+        if not(fut.finished()):
+          # Can't get strName anymore
+          const msg = "Async procedure yielded `nil`, " &
+                      "are you await'ing a `nil` Future?"
+          raiseAssert msg
+      else:
+        GC_ref(fut)
+        next.addCallback(internalContinue[T], cast[pointer](fut))
+  except CancelledError:
+    fut.cancelAndSchedule()
+  except CatchableError as exc:
+    #TODO what was that? futureVarCompletions
+    fut.fail(exc)
+  except Exception as exc:
+    if exc of Defect:
+      raise (ref Defect)(exc)
+
+    #TODO what was that? futureVarCompletions
+    fut.fail((ref ValueError)(msg: exc.msg, parent: exc))
+
+proc internalContinue[T](fut: pointer) {.gcsafe, raises: [Defect].} =
+  let asFut = cast[Future[T]](fut)
+  futureContinue(asFut)
+  GC_unref(asFut)
```

Comment:
It makes more sense to place the `GC_unref` call immediately after we restore the `asFut` variable.
This eliminates the risk of not reaching this line due to an exception from `futureContinue`

Reply:
From what I can see, the variable is never assigned in the generated C, so if we unref, it may be freed during the `futureContinue` because the refcount = 0

```c
N_LIB_PRIVATE N_NIMCALL(void, internalContinue__tOi1GGiApxKdBYtRcYwB7A)(void* fut) {
        tyObject_FuturecolonObjectType___GXFSekg1U8JRoedGa2vBSA* asFut;
        asFut = ((tyObject_FuturecolonObjectType___GXFSekg1U8JRoedGa2vBSA*) (fut));
        futureContinue__XuNTB7fHwBI8KII0qEQaCw_2(asFut);
        if (asFut) { nimGCunref(asFut); }
}
```

Reply:
It won't be freed because it exists on the stack. The conservative stack scanning will discover it.

Reply:
I'm pretty sure the stack scanning only happens during M&S. If you have a refcount = 0, it will be freed no matter what.

https://github.com/nim-lang/Nim/blob/0d0c249074d6a1041de16108dc247396efef5513/lib/system/gc.nim#L249-L250
Adds it to the ZCT if the refcount = 0, and the ZCT will be collected a bit later

I can test it, but it may be hard to catch that

Reply:
ZTC means "zero count table". This is where the objects start their life before they are assigned to a field of a heap allocated object and also where they end up once their ref count drops to zero. Most stack variables remain with ref count = 0 throughout their lifetime. Once an object is in the ZTC, it's a candidate for collection, but collection will only happen if a stack scan confirms that there are no remaining references to the object on the stack. 

This approach is known in the literature as "deferred reference counting" and it's what Nim's default GC employs:
https://en.wikipedia.org/wiki/Reference_counting#Dealing_with_inefficiency_of_updates

The goal is to avoid all the intermediate `addRef` and `decRef` calls while the stack variables are being passed around and copied. You count only the assignments to heap locations.

Reply:
Oh right, sorry, I saw all the stack scanning happening after the
`# ---------------- cycle collector`
line, so I imagined that stack scanning only happened during M&S

But indeed, the ZCT collection happens with the stack marked, so this should work, I'll try it out today :)

Reply:
It's indeed stable with GC_unref before continue, thanks for the explanation!